rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,295 @@
+
+During re-execution, child process will inherit all environment variables and
+flags.  `TELEPORT_TOOLS_VERSION=off` will be added during re-execution to
+prevent infinite loops.
+
+Automatic updates will not be used if `tctl` is connecting to the Auth Service
+over localhost.
+
+##### Errors and warnings
+
+If cluster administrator has chosen not to enroll client tools in automatic
+updates and does not self-manage client tools updates as outlined in
+[Self-managed client tools updates](#self-managed-client-tools-updates), a
+series of warnings and errors with increasing urgency will be shown to the
+user.
+
+If the version of client tools is within the same major version as advertised
+by the cluster, a warning will be shown to urge the user to enroll in automatic
+updates. Warnings will not prevent the user from using client tools that are
+slightly out of date.
+
+```
+$ tsh login --proxy=proxy.example.com
+Warning: Client tools are out of date, update to vX.Y.Z.
+
+Update Teleport to vX.Y.Z from https://goteleport.com/download or your system
+package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+
+[...]
+```
+
+If the version of client tools is 1 major version below the version advertised
+by the cluster, a warning will be shown that indicates some functionality may
+not work.
+
+```
+$ tsh login --proxy=proxy.example.com
+WARNING: Client tools are 1 major version out of date, update to vX.Y.Z.
+
+Some functionality may not work. Update Teleport to vX.Y.Z from
+https://goteleport.com/download or your system package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+```
+
+If the version of client tools is 2 (or more) versions lower than the version
+advertised by the cluster or 1 (or more) version greater than the version
+advertised by the cluster, an error will be shown and will require the user to
+use the `--skip-version-check` flag.
+
+```
+$ tsh login --proxy=proxy.example.com
+ERROR: Client tools are N major versions out of date, update to vX.Y.Z.
+
+Your cluster requires {tsh,tctl} vX.Y.Z. Update Teleport from
+https://goteleport.com/download or your system package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+
+Use the "--skip-version-check" flag to bypass this check and attempt to connect
+to this cluster.
+```
+
+#### Self-managed client tools updates
+
+Cluster administrators that want to self-manage client tools updates will be
+able to watch for changes to client tools versions which can then be used to
+trigger other integrations (using MDM software like JamF) to update the
+installed version of client tools on endpoints.
+
+```
+$ tctl autoupdate watch
+{"tools_version": "1.0.0"}
+{"tools_version": "1.0.1"}
+{"tools_version": "2.0.0"}
+
+[...]
+```
+
```

Comment:
Might be worth adding a `tctl autoupdate get` command as well, for integrating with polling systems that expect one-shot execution, like scripts.